eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala

```diff
@@ -228,43 +244,52 @@ class Peer(nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: Actor
       log.info(s"sent all announcements to {}: channels={} updates={} nodes={}", remoteNodeId, channelsSent, updatesSent, nodesSent)
       stay
 
-    case Event(Rebroadcast(channels, updates, nodes), ConnectedData(_, transport, _, _)) =>
-      // we filter out announcements that we received from this node
-      def sendIfNeeded(m: Map[_ <: LightningMessage, Set[ActorRef]]): Int = m.foldLeft(0) {
-        case (counter, (_, origins)) if origins.contains(self) =>
-          counter // won't send back announcement we received from this peer
-        case (counter, (announcement, _)) =>
-          transport ! announcement
-          counter + 1
-      }
-      val channelsSent = sendIfNeeded(channels)
-      val updatesSent = sendIfNeeded(updates)
-      val nodesSent = sendIfNeeded(nodes)
+    case Event(rebroadcast: Rebroadcast, ConnectedData(_, transport, _, _, maybeGossipTimestampFilter)) =>
+      val (channels1, updates1, nodes1) = Peer.filterGossipMessages(rebroadcast, self, maybeGossipTimestampFilter)
+      val channelsSent = channels1.length
```

Comment:
I'd rather do that in a `fold` (like it was before) so that we don't iterate twice

Reply:
we need first to process the updates, which we'll use to get a timestamp for channel announcements, and I could not yet find an elegant way to apply the fold pattern here but I've improved `filterGossipMessages` instead

Reply:
I think we can do slightly better than that:

```diff
diff --git a/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala b/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala
index 7f722bd5..28ada433 100644
--- a/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala
+++ b/eclair-core/src/main/scala/fr/acinq/eclair/io/Peer.scala
@@ -247,12 +247,20 @@ class Peer(nodeParams: NodeParams, remoteNodeId: PublicKey, authenticator: Actor
 
     case Event(rebroadcast: Rebroadcast, ConnectedData(_, transport, _, _, maybeGossipTimestampFilter)) =>
       val (channels1, updates1, nodes1) = Peer.filterGossipMessages(rebroadcast, self, maybeGossipTimestampFilter)
-      val channelsSent = channels1.length
-      val updatesSent = updates1.length
-      val nodesSent = nodes1.length
-      channels1.foreach(transport ! _)
-      updates1.foreach(transport ! _)
-      nodes1.foreach(transport ! _)
+
+      /**
+        * Send and count in a single iteration
+        */
+      def sendAndCount(msgs: Iterable[RoutingMessage]): Int = msgs.foldLeft(0) {
+        case (count, msg) =>
+          transport ! msg
+          count + 1
+      }
+
+      val channelsSent = sendAndCount(channels1)
+      val updatesSent = sendAndCount(updates1)
+      val nodesSent = sendAndCount(nodes1)
+
       if (channelsSent > 0 || updatesSent > 0 || nodesSent > 0) {
         log.debug(s"sent announcements to {}: channels={} updates={} nodes={}", remoteNodeId, channelsSent, updatesSent, nodesSent)
       }
@@ -436,7 +444,7 @@ object Peer {
     * @param gossipTimestampFilter optional gossip timestamp range
     * @return a filtered (channel announcements, channel updates, node announcements) tuple
     */
-  def filterGossipMessages(rebroadcast: Rebroadcast, self: ActorRef, gossipTimestampFilter: Option[GossipTimestampFilter]): (Seq[ChannelAnnouncement], Seq[ChannelUpdate], Seq[NodeAnnouncement]) = {
+  def filterGossipMessages(rebroadcast: Rebroadcast, self: ActorRef, gossipTimestampFilter: Option[GossipTimestampFilter]): (Iterable[ChannelAnnouncement], Iterable[ChannelUpdate], Iterable[NodeAnnouncement]) = {
 
     // check if this message has a timestamp that matches our timestamp filter
     def checkTimestamp(routingMessage: RoutingMessage): Boolean = gossipTimestampFilter match {
@@ -448,14 +456,14 @@ object Peer {
     }
 
     // we filter out announcements that we received from this node, and we also filter out updates against their timestamp filter
-    val updates1 = rebroadcast.updates.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a} toSeq
-    val nodes1 = rebroadcast.nodes.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a } toSeq
+    val updates1 = rebroadcast.updates.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a}
+    val nodes1 = rebroadcast.nodes.collect { case (a, origins) if !origins.contains(self) && checkTimestamp(a) => a }
 
     // keep a set of short channel ids for which we have an update
     val shortChannelIds = updates1.map(_.shortChannelId).toSet
 
     // filter out channels for which we don't have an update
-    val channels1 = rebroadcast.channels.collect { case (a, origins) if !origins.contains(self) && shortChannelIds.contains(a.shortChannelId) => a } toSeq
+    val channels1 = rebroadcast.channels.collect { case (a, origins) if !origins.contains(self) && shortChannelIds.contains(a.shortChannelId) => a }
 
     (channels1, updates1, nodes1)
   }
diff --git a/eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala b/eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala
index 98e755cf..93b6adba 100644
--- a/eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala
+++ b/eclair-core/src/test/scala/fr/acinq/eclair/io/PeerSpec.scala
@@ -51,6 +51,6 @@ class PeerSpec extends TestkitBaseClass {
     val (channels1, updates1, nodes1) = Peer.filterGossipMessages(rebroadcast, probe.ref, Some(GossipTimestampFilter(Block.RegtestGenesisBlock.blockId, timestamps.head, timestamps.last - timestamps.head)))
     assert(updates1.toSet == updates.filter(u => timestamps.contains(u.timestamp)).toSet)
     assert(nodes1.toSet == nodes.filter(u => timestamps.contains(u.timestamp)).toSet)
-    assert(channels1.toSet == channels.filter(ca => updates1.map(_.shortChannelId).contains(ca.shortChannelId)).toSet)
+    assert(channels1.toSet == channels.filter(ca => updates1.map(_.shortChannelId).toSet.contains(ca.shortChannelId)).toSet)
   }
 }

```